Adding ajax

Mikeysax opened the pull request:
The testing passes but according the the CI there is still failure. I've tried to add a `circle.yml` file to specify the node version to run against. The generator function is not recognized by the CI?

```
/home/ubuntu/otis-bot/test/grabURITest.coffee:17
        return function*() {
                       ^
SyntaxError: Unexpected token *
```

This is not code that exists. Coffee script is screwing up or am I missing a configuration on circleCI? @lasley @jonmast @radiofreemattd

Commit:
Added regex listener, added http post request with regex match

Commit:
Removed commented code

Commit:
Added hubot-test-helper and framework for 1 test

Commit:
Added to heroku, modifying tests

Commit:
Moved regex to variable, fixed test Outdent error for coffeescript

Commit:
Added basic testing suite

This commit adds in a basic testing infrastructure based on
Mocha.js and Chai.js, together with a helper framework
designed to test Hubot. Basic tests are included for the
Regular Expression used to to capture URLs and basic POST
functionality to the backend API.

Commit:
Modified for CI

This commit is intended to trigger push to CI server.

Commit:
Hubot parses URI properly with expected response

Otis hubot will parse the correct URL from chat and respond
accordingly.

Commit:
Fixed syntax inside if statement

Fixed the if statement syntax to include the error on the data.

Commit:
Removed extra required dep.

Removed Nock from testing.

Commit:
Modified testing to include co

Modified testing to include co because CI was failing even though local testing
indicates that everything is passing.

Commit:
Testing CI

Commit:
Testing CI

Commit:
testing CI

Commit:
Added circle.yml file

Commit:
testing CI

Commit:
Erased circle.yml

Commit:
Added circle.yml with node version 6.0.0

Commit:
Modified Readme, testing coveralls

Mikeysax commented:
Had to do with a configuration file for the CI, did not recognize the correct Node version. 🔨

NathanGE commented:
:thumbsup:

jonmast commented on test/grabURITest.coffee:
```diff
@@ -0,0 +1,20 @@
+  context 'User posts URL', ->
+    room = helper.createRoom()
+    co =>
+      yield room.user.say('user1', 'http://fakewebsite.com')
```
Forgive me if I'm just being stupid (JS is not my strong point), but what does wrapping this code with co do? I tried it without the wrapper and it seemed to work fine.

Mikeysax replied:
It's an artifact from all the problem solving of failure on the CI. I'll be removing it and refactoring and adding more tests.

Commit:
Fixed script if statement

Commit:
Added more tests

Added more regex and stubbing of http server with nock.
Currently, otis correctly parses URI.

Commit:
Added coveralls markdown badge to README.md

Commit:
Added istanbul for testing coverage

Refactored testing of Otis and added istanbul to account for coverage percent

Commit:
Fixing coverage commands

Commit:
Added coverage to gitignore and modified package.json

Mikeysax commented:
I think this is good. One of the biggest issues is getting coverage to write but that shouldn't affect anything overall.

I was getting this every time I run coverage: 

```
No coverage information was collected, exit without writing coverage information
```

I've tried a bunch of things, can't seem to fix the issue. Otherwise this should be good.

NathanGE commented:
I would agree that the coveralls report properly showing coverage isn't pivotal right now. It appears all previous concerns regarding test coverage and removing co have been addressed. I'm re-giving 👍.

lasley commented on test/regExpTest.coffee:
```diff
@@ -0,0 +1,71 @@
+expect = require('chai').expect
+
+describe 'regExp matches URI', ->
+  regExp = new RegExp("^(http[s]?:\\/\\/(www\\.)?|ftp:\\/\\/(www\\.)?|www\\.){1}([0-9A-Za-z-\\.@:%_\+~#=]+)+((\\.[a-zA-Z]{2,3})+)(/(.)*)?(\\?(.)*)?", "i")
```
This regex is incredibly complex, and doesn't even use things like `\d` and `\w` to make it easier to read. Why not simply:

```
[a-z]+[:.].*?(?=\s)
```

lasley replied:
If you leave the above, you need to test every part of it. Things not tested include support for username and passwords in the URI (user:pass@host), anchors(#), home path on ftp (~), etc. 

Also the `i` flag at the end indicates case insensitive, which means all the `[A-Za-z]` can simply be `[a-z]`.

Mikeysax replied:
That RegEx you posted is magical, hah. Witchcratch!

Mikeysax replied:
Previous Regex: 

```
Data {"url":"http://fakewebsite.com"}
```

Your Regex:

```
Data: {"url":"http://fakeweb"}
```

Your regex was working on other testing sites, but when I console logged the data it didn't contain the whole site URL

lasley replied:
If you can't get mine to work, I recommend simplifying your's or doubling up your tests to hit every character of the parse. This is maybe 50% tested as it is

lasley replied:
Your results don't make sense though, because I am using a negative lookahead to stop at a space. There isn't a space in the URI presented in your sample, so yeah doesn't make sense. Should look into the flags you are using in the regex

Mikeysax replied:
I've decided to use: https://github.com/kevva/url-regex

It seems to work well enough.

lasley replied:
Works for me

lasley commented on scripts/grabURI.coffee:
```diff
@@ -1,107 +1,21 @@
+# Place Holder apiUrl
+apiURL = "#"
+# Regular Expression for finding URI
+regExp = new RegExp("^(http[s]?:\\/\\/(www\\.)?|ftp:\\/\\/(www\\.)?|www\\.){1}([0-9A-Za-z-\\.@:%_\+~#=]+)+((\\.[a-zA-Z]{2,3})+)(/(.)*)?(\\?(.)*)?", "i")
```
Take a look below for regex advice. Also, why is the regex duped in the test?

lasley commented on scripts/grabURI.coffee:
```diff
@@ -1,107 +1,21 @@
+    robot.http(apiURL)
+      .header('Content-Type', 'application/json')
+      .post(data) (err, res, body) ->
+        if err
```
This looks untested

Mikeysax replied:
Do you have a tip for testing this? I'm not sure how to test AJAX calls?

lasley replied:
Return an error w/ nock instead of a 200

lasley commented:
Glad to see you figured out `nock`! 

I'm ok with not fixing coveralls, but make a ticket so someone does 😉

Mikeysax commented:
I don't feel like I figured out nock haha 🌵 but maybe I did 😖  lol

Commit:
Refactor of tests to use beforeEach and afterEach

Refactored all tests to be DRY. Imported urlRegex to match URL, works as
expected.

radiofreemattd commented:
It still seems like we are not testing the negative case (i.e. the err case).

If we can also eliminate any duplicate README files or anything similar, and make sure that a license file is added and an appropriate license header is added at the top of each file.

Commit:
Added MIT license and headers, added tests for post failure

radiofreemattd commented on test/grabURITest.coffee:
```diff
@@ -0,0 +1,205 @@
+        expect(room.messages).to.eql [
+          [ 'user1', 'http://fakewebsite.com' ]
+          [ 'hubot', "Meow, Uploading: http://fakewebsite.com" ]
+        ]
```
Shouldn't this expect that the error message is returned?

Mikeysax replied:
I don't think I can use Nock to test this. The HTTP server created by the Hubot Helper is self contained.

Mikeysax replied:
I'm trying to follow the way they test the server in the Helper package: https://github.com/mtsmfm/hubot-test-helper/blob/master/test/httpd-world_test.coffee

So far, I've not had any success: 

```
TypeError: http.post is not a function
```

Mikeysax replied:
@lasley @radiofreemattd I'm really not sure how to test this...it's confusing, could use some help.

matthewsdunn committed:
Revise and re-organize testing suite

This commit re-organizes the testing suite such that the post
functionality is completely separate from the testing of the
Regular Expression.

Still remaining is the testing of the negative case.

matthewsdunn committed:
Resolves all but the negative case test.

This commit would appear to capture all three messages, but there
is still a change appearing in the output that fails the expectation.

matthewsdunn committed:
Fix all tests.

This commit adds a fix for the negative case.

matthewsdunn committed:
Revise tests.

matthewsdunn committed:
Add in Bluebird promises and delay.

This commit implements Bluebird promises in order to avoid a
race condition that occurs between the posting of the message
to the stubbed Slack server and the evaluation of the expecation.

lasley commented on test/grabURITest.coffee:
```diff
@@ -0,0 +1,222 @@
+# Copyright (C) 2016 The Firehose Project
+#
+# This software may be modified and distributed under the terms
+# of the MIT license.  See the LICENSE file for details.
```
Better to point at a URI instead of a license file. The header is assuming that the file has been broken out of context, so the license file is likely not there. Below is my std header for MIT:

```
# © YEAR LasLabs Inc.
# License MIT (https://opensource.org/licenses/MIT).
```

lasley replied:
Note how it's short and sweet. Screw the whitespace and extra lines, the shorter the better. You need to enforce copyright and license, everything else is bust.

lasley commented on scripts/grabURI.coffee:
```diff
@@ -1,107 +1,28 @@
 # Description:
```
If we're leaving this blank, we should just kill the lines. Also, copyright/license header is missing

lasley commented on MITLicense.md:
```diff
@@ -0,0 +1,8 @@
+The MIT License (MIT)
```
This isn't actually Markdown. Recommend switching file format to txt or using Markdown version - https://github.com/IQAndreas/markdown-licenses/blob/master/mit.md

lasley commented on README.md:
```diff
@@ -1,214 +1 @@
-
-You may want to get comfortable with `heroku logs` and `heroku restart` if
-you're having issues.
+[![Coverage Status](https://coveralls.io/repos/github/otis-bot/otis-bot/badge.svg?branch=master)](https://coveralls.io/github/otis-bot/otis-bot?branch=master)
```
CI status?

lasley replied:
BTW here's a template I use that may help here - https://repo.laslabs.com/projects/TEM/repos/odoo-module_template/browse/README.md?at=daf9831fedad5fa7488dcc82e5ca480c3d137ad4&raw

matthewsdunn committed:
Added license headers

This commit adds license information throughout the files
created by the team and additionally tried to make more
uniform the coding style of the project.

matthewsdunn committed:
Add coffeelint as devlopment dependency

This commit adds coffeelint as a development dependency to the
package.json of the project.

matthewsdunn committed:
Configure package.json for linting

This commit adds a new coffescript linting task to the package.json

matthewsdunn committed:
Modify test file to accomodate linter

This commit consists of a series of changes requested by the
coffeescript linter.

matthewsdunn committed:
Remove unnecessary files

This commit removes unnecessary files left from the basic
installation of the Hubot boilerplate code.

Mikeysax commented:
I think we should keep those dependencies that were just removed, otherwise 👍

Commit:
Readded robot.emit to prevent hubot crash

Added robot.emit to prevent hubot from crashing and also recommended by
hubot documentation.

Commit:
Added npm-debug.log to gitignore

Commit:
Testing linter and added link to readme

Testing lint command on circleCI and added my name and github link on
readme under Matt. Corrected some typos in the README.md.

Commit:
Fixed typos

lasley commented:
LGTM, good job guys 👍

Mikeysax commented:
I think this is good :+1: merge?
